diffusion/models/stable_diffusion.py

```diff
@@ -142,52 +143,52 @@ def set_rng_generator(self, rng_generator: torch.Generator):
         if self.precomputed_latents and self.image_latents_key in batch and self.text_latents_key in batch:
             if self.sdxl:
                 raise NotImplementedError('SDXL not yet supported with precomputed latents')
-            latents, conditioning = batch[self.image_latents_key], batch[self.text_latents_key]
+            latents, text_embeds = batch[self.image_latents_key], batch[self.text_latents_key]
         else:
-            inputs, conditioning = batch[self.image_key], batch[self.text_key]
-            if self.sdxl:
-                # If SDXL, separate the conditioning ([B, 2, 77]) from each tokenizer
-                conditioning, conditioning_2 = conditioning[:, 0, :], conditioning[:, 1, :]
-                conditioning_2 = conditioning_2.view(-1, conditioning_2.shape[-1])
-
-            conditioning = conditioning.view(-1, conditioning.shape[-1])
-            if self.encode_latents_in_fp16:
-                # Disable autocast context as models are in fp16
-                with torch.cuda.amp.autocast(enabled=False):
-                    # Encode the images to the latent space.
-                    # Encode prompt into conditioning vector
-                    latents = self.vae.encode(inputs.half())['latent_dist'].sample().data
-                    if self.sdxl:
-                        conditioning, pooled_conditioning = self.text_encoder([conditioning, conditioning_2])
-                    else:
-                        conditioning = self.text_encoder(conditioning)[0]  # Should be (batch_size, 77, 768)
+            inputs, conditionings = batch[self.image_key], batch[self.text_key]
 
-            else:
-                latents = self.vae.encode(inputs)['latent_dist'].sample().data
-                if self.sdxl:
-                    conditioning, pooled_conditioning = self.text_encoder([conditioning, conditioning_2])
+            # If encode_latents_in_fp16, disable autocast context as models are in fp16
+            c = torch.cuda.amp.autocast(enabled=False) if self.encode_latents_in_fp16 else nullcontext()  # type: ignore
+            with c:
+                # Encode the images to the latent space.
+                if self.encode_latents_in_fp16:
+                    latents = self.vae.encode(inputs.half())['latent_dist'].sample().data
                 else:
-                    conditioning = self.text_encoder(conditioning)[0]
+                    latents = self.vae.encode(inputs)['latent_dist'].sample().data
+                # Encode tokenized prompt into embedded text and pooled text embeddings
+                text_encoder_out = self.text_encoder(conditionings, attention_mask=attention_mask)
+                text_embeds = text_encoder_out[0]
+                if self.sdxl:
+                    if len(text_encoder_out) <= 1:
+                        raise RuntimeError('SDXL requires text encoder output to include a pooled text embedding')
+                    text_pooled_embeds = text_encoder_out[1]
```

Comment:
The safest option would be to specify a key to index for the text embeddings and the projection embedding, but this would add two more arguments to the model and this PR is already pretty long...